### PR TITLE
ci: Optimize build + test pipeline

### PR DESCRIPTION
**Pull request by frascuchon**
# Description

This PR reviews and organizes the CI pipeline. With the changes introduced here, all tests are running as expected.  

Also, the cache action has been updated in order to fix a cache resolution problem which allows to reuse the cache for the rest of the jobs.

## Two blocks of tests are launched separately 

**base tests** (most of the server tests and some basic client tests) and **extra tests** (training, monitor, listeners...). The pipeline falls into errors if resource all tests are launched in the same job. 
In the next step, those groups should be reviewed and improved.  

Refs #3291 

## Reusable jobs extraction

The job running the tests has been extracted as a separate workflow, which helps with reusability and workflow code maintenance. The same has been applied to the build Python package.

## Improve conditional executions

The `IS_DEPLOYABLE` step conditions have been removed. Now, this condition is executed just once at the related job level.

### CODE REVIEW

File: .github/workflows/build-python-package.yml [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+      - name: Cache pip 👜
+        uses: actions/cache@v3
+        env:
+          # Increase this value to reset cache if pyproject.toml has not changed
+          CACHE_NUMBER: 0
+        with:
+          path: ~/.cache/pip
+          key: ${{ runner.os }}-pip-${{ env.CACHE_NUMBER }}-${{ hashFiles('pyproject.toml') }}
+      - name: Setup Node.js
+        uses: actions/setup-node@v2
```
Comment (gabrielmbmb) [inline .github/workflows/build-python-package.yml]: should we update this to `actions/setup-node@v3`?

File: .github/workflows/build-python-package.yml [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+        with:
+          node-version: "14"
+      - name: Build Package 🍟
+        run: |
+          pip install -U build
+          scripts/build_distribution.sh
+      - name: Upload package artifact
+        uses: actions/upload-artifact@v2
```
Comment (gabrielmbmb) [inline .github/workflows/build-python-package.yml]: should we update this to `actions/upload-artifact@v3`?

File: .github/workflows/run-python-tests.yml [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+      - name: Setup ElasticSearch 🔎
+        if: inputs.searchEngine == 'elasticsearch'
+        uses: ankane/setup-elasticsearch@v1
+        with:
+          elasticsearch-version: ${{ inputs.searchEngineVersion }}
+      - name: Setup OpenSearch 🔎
+        if: inputs.searchEngine == 'opensearch'
+        uses: ankane/setup-opensearch@v1
+        with:
+          opensearch-version: ${{ inputs.searchEngineVersion }}
```
Comment (gabrielmbmb) [inline .github/workflows/run-python-tests.yml:82]: maybe we could replace these actions with this https://docs.github.com/en/actions/using-containerized-services/about-service-containers#creating-service-containers?

**Review by gabrielmbmb — APPROVED** [review]
LGTM!

File: .github/workflows/run-python-tests.yml (hunk 3 above)
Comment (frascuchon) [reply .github/workflows/run-python-tests.yml:82]: Yes, for sure. I will tackle this in a separate PR. Thanks!

File: .github/workflows/build-python-package.yml (hunk 2 above)
Comment (frascuchon) [reply .github/workflows/build-python-package.yml]: Done

File: .github/workflows/build-python-package.yml (hunk 1 above)
Comment (frascuchon) [reply .github/workflows/build-python-package.yml]: Done

File: .github/workflows/run-python-tests.yml [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+          pip install -e ".[server,listeners]"
+          alembic upgrade head
+          pytest --cov=argilla --cov-report=xml ${{ inputs.pytestArgs }}
+      - name: Upload Coverage to Codecov 📦
+        uses: codecov/codecov-action@v1
+        if: inputs.coverageReport
+        with:
+          files: ./coverage.xml
+          flags: pytest
+          fail_ci_if_error: false
```
Comment (gabrielmbmb) [inline .github/workflows/run-python-tests.yml]: I think that we should remove this from here, and have an step in `package.yml` that [combines](https://coverage.readthedocs.io/en/stable/cmd.html#combining-data-files-coverage-combine) the `coverage` files from both jobs (`run_tests` and `run_tests_extra`) and then upload the combined file to codecov.

File: pyproject.toml [hunk 5]
```diff
@@ -131,7 +131,7 @@ log_format = "%(asctime)s %(name)s %(levelname)s %(message)s"
 log_date_format = "%Y-%m-%d %H:%M:%S"
 log_cli = "True"
 testpaths = ["tests"]
-addopts = ["--ignore", "tests/training"]
```
Comment (tomaarsen) [inline pyproject.toml:134]: Nice, it seems good to run these tests locally too.

**Review by tomaarsen — APPROVED** [review]

File: .github/workflows/run-python-tests.yml (hunk 4 above)
Comment (frascuchon) [reply .github/workflows/run-python-tests.yml]: I finally could work on this with success. Now all coverage reports are combined and published automatically. So final coverage increased substantially

**Pull request merged**